src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -352,20 +486,29 @@ static async ValueTask<FlushResult> WriteDataAndTrailersAsyncCore(Http2FrameWrit
         |                           Padding (*)                       ...
         +---------------------------------------------------------------+
     */
-    private void WriteDataUnsynchronized(Http2Stream stream, in ReadOnlySequence<byte> data, long dataLength, bool endStream)
+    private void WriteDataUnsynchronized(int streamId, in ReadOnlySequence<byte> data, long dataLength, bool endStream)
     {
         Debug.Assert(dataLength == data.Length);
 
         // Note padding is not implemented
-        _outgoingFrame.PrepareData(stream.StreamId);
+        _outgoingFrame.PrepareData(streamId);
 
         if (dataLength > _maxFrameSize) // Minus padding
         {
             TrimAndWriteDataUnsynchronized(in data, dataLength, endStream);
             return;
         }
 
-        WriteDataUnsynchronizedCore(stream, endStream, dataLength, data);
+        if (endStream)
+        {
+            _outgoingFrame.DataFlags |= Http2DataFrameFlags.END_STREAM;
+        }
+
+        _outgoingFrame.PayloadLength = (int)dataLength; // Plus padding
+
+        WriteHeaderUnsynchronized();
+
+        data.CopyTo(_outputWriter);
```

Comment:
Why the inconsistency between `CopyTo` in some places an foreaching over the data in others?

Reply:
This is unchanged codez.

Reply:
Change it to CopyTo everywhere. It has an optimized path for single segment ROS.